### PR TITLE
[Chat] Rlhf support SimPO

### PR DESCRIPTION
**Pull request by YeAnbang**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [ ] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs
- [x] I have installed pre-commit: `pip install pre-commit && pre-commit install`

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: applications/ColossalChat/examples/README.md [hunk 1]
```diff
@@ -717,14 +718,29 @@ For DPO training, you only need the preference dataset. Please follow the instru
...
 #### Step 2: Training
-You can run the [train_dpo.sh](./examples/training_scripts/train_dpo.sh) to start DPO training. Please refer to the [training configuration](#training-configuration) section for details regarding supported training options.
+You can run the [train_dpo.sh](./examples/training_scripts/train_dpo.sh) to start DPO training. Please refer to the [training configuration](#training-configuration) section for details regarding supported training options. Following the trend of recent research on DPO-like alignment methods, we added option for the user to choose from, including whether to do length normalization , reward shaping and whether to use a reference model in calculating implicit reward. Here are those options,
 
+```
+--beta 0.1 \     # the temperature in DPO loss, Default to 0.1
+--gamma 0.0 \     # the reward target margin in the SimPO paper, Default to 0.
+--disable_reference_model \   # whether to disable the reference model, if set, the implicit reward will be calculated solely from the actor. Default to enable reference model in DPO
+--length_normalization \  # whether to apply length normalization, Default to not use
+```
```
Comment (TongLi3701) [inline applications/ColossalChat/examples/README.md:728]: I suggest refactoring the SimPO part to a separate loss to make it more clearer. I understand both are preference optimization method so you can share the same DPOTrainer but make it more clearer for different method. For example, add one option for POloss ("dpo" or "simpo"). 

User's may not have clear idea of using parameters here, for example, `disable_reference_model` can trigger SimPO at the backend. Thus, we should clearly explain if you want to use DPO, we can adjust which parameters, if you want to use SimPO, you can use which parameters.

File: applications/ColossalChat/examples/README.md [hunk 2]
```diff
@@ -717,14 +718,29 @@ For DPO training, you only need the preference dataset. Please follow the instru
...
 
+### Alternative Option For RLHF: Simple Preference Optimization
+
+We support the method introduced in the paper [SimPO: Simple Preference Optimization
+with a Reference-Free Reward](https://arxiv.org/pdf/2405.14734) (SimPO). Which is a reference model free aligment method that add length normalization and reward shaping to the DPO loss to enhance training stability and efficiency. As the method doesn't deviate too much from DPO, we add support for length normalization and SimPO reward shaping in our DPO implementation. Simply set the flag to disable the use of the reference model, set the reward target margin and enable length normalization in the DPO training script.
+
+#### SimPO Result
+<p align="center">
+<img width="1000" alt="image" src="https://raw.githubusercontent.com/hpcaitech/public_assets/main/applications/chat/SimPO_margin.png">
+</p>
```
Comment (TongLi3701) [inline applications/ColossalChat/examples/README.md:743]: We should make the explanation of SimPO related parameters here.

Comment (TongLi3701) [reply applications/ColossalChat/examples/README.md:743]: For DPO-like method such as "orpo", you can keep the same structure.

**Review by TongLi3701 — CHANGES_REQUESTED** [review]
Thanks Anbang, I left some comments. Please have a look.

File: applications/ColossalChat/examples/README.md (hunk 2 above)
Comment (YeAnbang) [reply applications/ColossalChat/examples/README.md:743]: orpo is not the same. will create a separate file

File: applications/ColossalChat/coati/dataset/loader.py [hunk 3]
```diff
@@ -187,6 +187,14 @@ def __call__(self, instances: Sequence[Dict[str, List[int]]]) -> Dict[str, torch
...
         )
 
+        torch.set_printoptions(profile="full")
+
+        for ins in instances:
+            if sum(ins["chosen_loss_mask"][1:]) == 0:
+                print("Before truncated", ins["chosen_loss_mask"], len(ins["chosen_loss_mask"]))
+            if sum(ins["rejected_loss_mask"][1:]) == 0:
+                print("Before truncated", ins["rejected_loss_mask"], len(ins["rejected_loss_mask"]))
+
```
[the remark this answers is not in the document]
Comment (YeAnbang) [reply applications/ColossalChat/coati/dataset/loader.py:197]: will remove those

File: applications/ColossalChat/examples/README.md (hunk 1 above)
Comment (YeAnbang) [reply applications/ColossalChat/examples/README.md:728]: will do

File: applications/ColossalChat/examples/training_scripts/train_dpo.py [hunk 4]
```diff
@@ -128,7 +128,7 @@ def train(args):
             disable_dropout(ref_model)
         else:
             ref_model = None
-
+        print("ref_model is None", args.disable_reference_model, ref_model is None)
```
Comment (TongLi3701) [inline applications/ColossalChat/examples/training_scripts/train_dpo.py:131]: Use logger rather than print

**Review by TongLi3701 — APPROVED** [review]
Thanks, Anbang. Please remove TODO list in the README.

I left some comments. Please address them and merge.

**Pull request merged**